crates/rustc_codegen_spirv/src/builder_spirv.rs

```diff
@@ -329,44 +414,143 @@ impl BuilderSpirv {
-        if let Some(value) = self.constants.borrow_mut().get_by_left(&val) {
-            return *value;
+        if let Some(entry) = self.const_to_id.borrow().get(&val_with_type) {
+            // FIXME(eddyb) deduplicate this `if`-`else` and its other copies.
+            let kind = if entry.legal.is_ok() {
+                SpirvValueKind::Def(entry.val)
+            } else {
+                SpirvValueKind::IllegalConst(entry.val)
+            };
+            return SpirvValue { kind, ty };
         }
+        let val = val_with_type.val;
         let id = match val {
-            SpirvConst::U32(ty, v) => builder.constant_u32(ty, v).with_type(ty),
-            SpirvConst::U64(ty, v) => builder.constant_u64(ty, v).with_type(ty),
-            SpirvConst::F32(ty, v) => builder.constant_f32(ty, f32::from_bits(v)).with_type(ty),
-            SpirvConst::F64(ty, v) => builder.constant_f64(ty, f64::from_bits(v)).with_type(ty),
-            SpirvConst::Bool(ty, v) => {
+            SpirvConst::U32(v) => builder.constant_u32(ty, v),
+            SpirvConst::U64(v) => builder.constant_u64(ty, v),
+            SpirvConst::F32(v) => builder.constant_f32(ty, f32::from_bits(v)),
+            SpirvConst::F64(v) => builder.constant_f64(ty, f64::from_bits(v)),
+            SpirvConst::Bool(v) => {
                 if v {
-                    builder.constant_true(ty).with_type(ty)
+                    builder.constant_true(ty)
                 } else {
-                    builder.constant_false(ty).with_type(ty)
+                    builder.constant_false(ty)
                 }
             }
-            SpirvConst::Composite(ty, ref v) => builder
-                .constant_composite(ty, v.iter().copied())
-                .with_type(ty),
-            SpirvConst::Null(ty) => builder.constant_null(ty).with_type(ty),
-            SpirvConst::Undef(ty) => builder.undef(ty, None).with_type(ty),
+
+            SpirvConst::Null => builder.constant_null(ty),
+            SpirvConst::Undef | SpirvConst::ZombieUndefForFnAddr => builder.undef(ty, None),
+
+            SpirvConst::Composite(ref v) => builder.constant_composite(ty, v.iter().copied()),
+
+            SpirvConst::PtrTo { pointee } => {
+                builder.variable(ty, None, StorageClass::Private, Some(pointee))
+            }
         };
-        self.constants
-            .borrow_mut()
-            .insert_no_overwrite(val, id)
-            .unwrap();
-        id
+        #[allow(clippy::match_same_arms)]
+        let legal = match val {
+            SpirvConst::U32(_)
+            | SpirvConst::U64(_)
+            | SpirvConst::F32(_)
+            | SpirvConst::F64(_)
+            | SpirvConst::Bool(_) => Ok(()),
+
+            SpirvConst::Null => {
+                // FIXME(eddyb) check that the type supports `OpConstantNull`.
+                Ok(())
+            }
+            SpirvConst::Undef => {
+                // FIXME(eddyb) check that the type supports `OpUndef`.
```

Comment:
Just curious, are there any types that _cannot_ be made by OpUndef? Intuitively, there absolutely should be, but I don't think the spec actually mentions anything.